Remove sourcing of .env

chmurph2 opened the pull request:
This breaks how Dotenv sources .env.* files in our Rails apps.  Dotenv will first source any .env.local, .env.development, or .env.test files before sourcing .env.  Puma-dev sources .env before Dotenv is able to source as described above, so enviroment variables found in .env.* don't "override" the same ones found in .env.

This commit is a hack to workaround this issue, and experiment with providing our own homebrew tap. We will consider opening a proper pull request to Puma-dev if there's a better approach, such as a command-line flag that'll skip all the env sourcing going on here.

Hunter provided a binary build of this PR.

pboling commented:
@chmurph2 Does puma-dev work with dotenv?  I haven't been able to figure it out yet.

chmurph2 commented:
Hi @pboling. Does your approach allow environment variables in `.env.local` or `.env.development`  to override the same ones found in `.env`?

pboling commented:
Yes, `dotenv` is the executable hammer that ships with `dotenv` gem, to allow any shell command to inherit the dotenv variables, as configured, with overrides, as expected.

Adding `dotenv` to `.powrc`, `.powenv`, or any of the 4 magic script files that puma-dev looks for, will work.

chmurph2 commented:
OK, this is very good to know. Thank you!

lowski commented:
@chmurph2 did it work for you? I can’t check it now but I think it won’t work since puma dev simple source the “.powrc”, “powenv” flies (bash) while dotenv CLI requires to prefix command with it (dotenv ./script.rb)

chmurph2 commented:
Hi @lowski, are you asking if this PR works (it does), or if I tried pboling's suggestion above (I have not, yet).

lowski commented:
@chmurph2 thanks for reply. I was asking if @pboling’s solution works for you.

chmurph2 commented:
@lowski I just tested this and @pboling's suggestion above does _not_ work (though it was a good idea). Here's what I did:

1. Removed my company's fork of puma-dev and purged it from my system (involved rebooting)
1. Made sure the latest version of puma-dev is installed and working
1. Confirmed again that puma-dev breaks dotenv's [hierarchy](https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use), by adding overrides of env vars found in `.env` to `.env.development` in my app
1. Added `dotenv` to `.powrc` (and `.powenv` after in another test) in my app's root directory
1. The overrides in `.env.development` weren't honored, because (I presume) [puma-dev sources](https://github.com/puma/puma-dev/blob/master/dev/app.go#L242-L244) `.env` before `.powrc` and `.powenv`, and calling `dotenv` afterward in `.powrc` does nothing to help this.

Another workaround is to stop using `.env` and instead use `.env.development`, `.env.development.local`, `.env.test`, `.env.production`, etc., but that may not be possible for some. We'd certainly have to update a lot of our apps accordingly.

HTH,
Chris

lowski commented:
@chmurph2 thank you very much for checking. That's what I suspected.

I think there are 2 issues:

- Puma dev sources `.env` file so `dotenv` will not assign env variables according to its hierarchy because initial values (from `.env` file) were assigned.
- Running `dotenv` via `.powrc` file will not affect the process running the server, as variables will be assigned (visible) only for the `dotenv` process.

chmurph2 commented:
I believe so, yes.

chmurph2 commented:
Some follow up thoughts and recommendations; I believe the choices for puma-dev + dotenv users are as follows:
1. Use vanilla puma-dev and stop using `.env` and move the env vars (with sensible defaults) to `.env.development` and `.env.test`, and check those files in. Override them with .gitignored `.env.development.local` and `.env.test.local` as needed: https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use. Also, since `env.development` is read in before `.env.test`, one can effectively use `.env.development` with sensible env var defaults and only override specific ones in `.env.test` (if any).
2. Use vanilla puma-dev and change your app to use  `Dotenv.overload` for dev/test: https://github.com/bkeepers/dotenv#why-is-it-not-overriding-existing-env-variables
3. Use this fork of puma-dev. I no longer recommend this, as I assume it’s a matter of time until it stops working from some future update to MacOS libraries, etc..

pboling commented:
@chmurph2 was `3.` supposed to have a link to a fork?  Or are you referring to this project `puma/puma-dev` as a fork?

Thanks so much for documenting the first two options! ❤️

chmurph2 commented:
@pboling fixed. I was thinking of this closed PR as the "fork" at the time, but linking to our actual release is better.

lowski commented:
@chmurph2 thank you for update. One thing to keep in mind: `Dotenv.overload` does not automatically include `.env.test` / `env.development` files.
